File: gams/apifiles/Java/Warehouse.java
package com.gams.examples;

import java.util.HashMap;
import java.util.Map;

import com.gams.api.GAMSException;
import com.gams.api.GAMSExecutionException;
import com.gams.api.GAMSGlobals;
import com.gams.api.GAMSDatabase;
import com.gams.api.GAMSJob;
import com.gams.api.GAMSOptions;
import com.gams.api.GAMSVariableRecord;
import com.gams.api.GAMSWorkspace;

public class Warehouse {

    static int numberOfThreads = 16;
    static int status = 0;
    
    public static void main(String[] args) {

        GAMSWorkspace ws = new GAMSWorkspace();
        
        // create a GAMSDatabase for the results
        GAMSDatabase resultDB = ws.addDatabase();
        resultDB.addParameter("objrep",1,"Objective value");
        resultDB.addSet("supplyMap",3,"Supply connection with level");

        try {
            // run multiple parallel jobs
            Object dbLock = new Object();
            Map<String, WarehouseThread> warehousesMap = new HashMap<String, WarehouseThread>();
            for (int i=10; i<=numberOfThreads; i++) { 
                 WarehouseThread wh = new WarehouseThread(ws, i, resultDB, dbLock);
                 warehousesMap.put(Integer.toString(i), wh);
                 wh.start();
            }
        
            // join all threads
            for (WarehouseThread wh : warehousesMap.values()) {  
                 try {
                     wh.join();
                 } catch (InterruptedException e) {
                    e.printStackTrace();
                 }
            }
        
            // export the result database to a GDX file
            resultDB.export(ws.workingDirectory() + GAMSGlobals.FILE_SEPARATOR + "result.gdx");
        } catch (Exception e) {
             e.printStackTrace();
        } finally {
             resultDB.dispose();
        }
        
        System.exit(status);
    }
    
    public static void notifyException(GAMSException e) {
        if (e instanceof GAMSExecutionException) 
            status = ((GAMSExecutionException)e).getExitCode();
        else 
            status = -1;
    }
    
    static class WarehouseThread extends Thread {
        GAMSWorkspace workspace;
        GAMSDatabase result;
        Object lockObject;
        int numberOfWarehouses;
   
        public WarehouseThread(GAMSWorkspace ws, int number, GAMSDatabase db, Object lockObj) {
              workspace = ws;
              numberOfWarehouses = number;
              result = db;
              lockObject = lockObj;
        }
  
        public void run() {
            try {
               // instantiate GAMSOptions and define some scalars
               GAMSOptions opt = workspace.addOptions();
               opt.setAllModelTypes( "Gurobi" );
               opt.defines("Warehouse", Integer.toString(numberOfWarehouses));
               opt.defines("Store", "65");
               opt.defines("fixed", "22");
               opt.defines("disaggregate", "0");
               opt.setOptCR( 0.0 ); // Solve to optimality

               // create a GAMSJob from string and write results to the result database
               GAMSJob job = workspace.addJobFromString(model);
               job.run(opt, System.out);   // job.run(opt);

               // need to lock database write operations
               synchronized (lockObject) {
                   result.getParameter("objrep").addRecord(Integer.toString(numberOfWarehouses)).setValue( job.OutDB().getVariable("obj").findRecord().getLevel() );
               }   
                
               for (GAMSVariableRecord supplyRec : job.OutDB().getVariable("supply")) {
                 if (supplyRec.getLevel() > 0.5)
                    synchronized (lockObject) {
                         String[] keys = new String[] { Integer.toString(numberOfWarehouses), supplyRec.getKeys()[0], supplyRec.getKeys()[1] };
                         result.getSet("supplyMap").addRecord( keys );
                    }
              }
            } catch (GAMSException e) {
                e.printStackTrace();
                Warehouse.notifyException(e) ;
            }
        }
    }
    
    static String model =  
            "$title Warehouse.gms                                                      \n" +
            "                                                                          \n" +
            "$eolcom //                                                                \n" +
            "$SetDDList warehouse store fixed disaggregate // acceptable defines       \n" +
            "$if not set warehouse    $set warehouse   10                              \n" +
            "$if not set store        $set store       50                              \n" +  
            "$if not set fixed        $set fixed       20                              \n" +
            "$if not set disaggregate $set disaggregate 1 // indicator for tighter bigM constraint \n" +
            "$ife %store%<=%warehouse% $abort Increase number of stores (>%warehouse)  \n" +
            "                                                                          \n" +
            "Sets Warehouse  /w1*w%warehouse% /                                        \n" +  
            "     Store      /s1*s%store%     /                                        \n" +
            "Alias (Warehouse,w), (Store,s);                                           \n" +
            "Scalar                                                                    \n" +
            "     fixed        fixed cost for opening a warehouse / %fixed% /          \n" +
            "Parameter                                                                 \n" +
            "     capacity(WareHouse)                                                  \n" +
            "     supplyCost(Store,Warehouse);                                         \n" +
            "                                                                          \n" +
            "$eval storeDIVwarehouse trunc(card(store)/card(warehouse))                \n" + 
            "capacity(w)     =   %storeDIVwarehouse% + mod(ord(w),%storeDIVwarehouse%);\n" +
            "supplyCost(s,w) = 1+mod(ord(s)+10*ord(w), 100);                           \n" +
            "                                                                          \n" +
            "Variables                                                                 \n" +
            "    open(Warehouse)                                                       \n" +      
            "    supply(Store,Warehouse)                                               \n" +
            "    obj;                                                                  \n" +    
            "Binary variables open, supply;                                            \n" +
            "                                                                          \n" +
            "Equations                                                                 \n" +
            "    defobj                                                                \n" +  
            "    oneWarehouse(s)                                                       \n" +
            "    defopen(w);                                                           \n" +   
            "                                                                          \n" +
            "defobj..  obj =e= sum(w, fixed*open(w)) + sum((w,s), supplyCost(s,w)*supply(s,w));  \n" +
            "                                                                          \n" +
            "oneWarehouse(s).. sum(w, supply(s,w)) =e= 1;                              \n" +
            "                                                                          \n" +
            "defopen(w)..      sum(s, supply(s,w)) =l= open(w)*capacity(w);            \n" +
            "                                                                          \n" +
            "$ifthen %disaggregate%==1                                                 \n" +   
            "Equations                                                                 \n" +  
            "     defopen2(s,w);                                                       \n" +
            "defopen2(s,w).. supply(s,w) =l= open(w);                                  \n" +  
            "$endif                                                                    \n" +
            "                                                                          \n" +
            "model distrib /all/;                                                      \n" +
            "solve distrib min obj using mip;                                          \n" + 
            "abort$(distrib.solvestat<>%SolveStat.NormalCompletion% or                 \n" +   
            "       distrib.modelstat<>%ModelStat.Optimal% and                         \n" +
            "       distrib.modelstat<>%ModelStat.IntegerSolution%) 'No solution!';    \n" +
            "                                                                          \n";
}

File: gams/apifiles/Python/warehouse.py
from gams import *
import threading
import sys
import os

def get_model_text():
    return '''
$title Warehouse.gms

$eolcom //
$SetDDList warehouse store fixed disaggregate // acceptable defines
$if not set warehouse    $set warehouse   10
$if not set store        $set store       50
$if not set fixed        $set fixed       20
$if not set disaggregate $set disaggregate 1 // indicator for tighter bigM constraint
$ife %store%<=%warehouse% $abort Increase number of stores (>%warehouse)

Sets Warehouse  /w1*w%warehouse% /
     Store      /s1*s%store%     /
Alias (Warehouse,w), (Store,s);
Scalar
     fixed        fixed cost for opening a warehouse / %fixed% /
Parameter
     capacity(WareHouse)
     supplyCost(Store,Warehouse);

$eval storeDIVwarehouse trunc(card(store)/card(warehouse))
capacity(w)     =   %storeDIVwarehouse% + mod(ord(w),%storeDIVwarehouse%);
supplyCost(s,w) = 1+mod(ord(s)+10*ord(w), 100);

Variables
    open(Warehouse)
    supply(Store,Warehouse)
    obj;
Binary variables open, supply;

Equations
    defobj
    oneWarehouse(s)
    defopen(w);

defobj..  obj =e= sum(w, fixed*open(w)) + sum((w,s), supplyCost(s,w)*supply(s,w));

oneWarehouse(s).. sum(w, supply(s,w)) =e= 1;

defopen(w)..      sum(s, supply(s,w)) =l= open(w)*capacity(w);

$ifthen %disaggregate%==1
Equations
     defopen2(s,w);
defopen2(s,w).. supply(s,w) =l= open(w);
$endif

model distrib /all/;
solve distrib min obj using mip;
abort$(distrib.solvestat<>%SolveStat.NormalCompletion% or
       distrib.modelstat<>%ModelStat.Optimal% and
       distrib.modelstat<>%ModelStat.IntegerSolution%) 'No solution!'; '''


def solve_warehouse(workspace, number_of_warehouses, result, db_lock):
    try:
        # instantiate GAMSOptions and define some scalars
        opt = workspace.add_options()
        opt.all_model_types = "Gurobi"
        opt.defines["Warehouse"] = str(number_of_warehouses)
        opt.defines["Store"] = "65"
        opt.defines["fixed"] = "22"
        opt.defines["disaggregate"] = "0"
        opt.opt_cr = 0.0 # Solve to optimality

        # create a GAMSJob from string and write results to the result database
        job = workspace.add_job_from_string(get_model_text())
        job.run(opt)

        # need to lock database write operations
        db_lock.acquire()
        result["objrep"].add_record(str(number_of_warehouses)).value = job.out_db["obj"][()].level
        db_lock.release()
        for supply_rec in job.out_db["supply"]:
            if supply_rec.level > 0.5:
                db_lock.acquire()
                result["supplyMap"].add_record((str(number_of_warehouses), supply_rec.keys[0], supply_rec.keys[1]))
                db_lock.release()
    except GamsException as e:
        print "GamsException occured: " + str(e)
        os._exit(1)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        ws = GamsWorkspace(system_directory = sys.argv[1])
    else:
        ws = GamsWorkspace()
    
    # create a GAMSDatabase for the results
    result_db = ws.add_database()
    result_db.add_parameter("objrep" ,1 ,"Objective value")
    result_db.add_set("supplyMap" ,3 ,"Supply connection with level")
    
    try:
        # run multiple parallel jobs
        db_lock = threading.Lock()
        threads = {}
        for i in range(10,22):
            threads[i] = threading.Thread(target=solve_warehouse, args=(ws, i, result_db, db_lock))
            threads[i].start()
        for t in threads.values():
            t.join()
        
        # export the result database to a GDX file
        result_db.export("/tmp/result.gdx")
    
    except GamsException as e:
        print "GamsException occured: " , e
    except Exception as e:
        print e

 


